File: src/services/reportService.ts
/**
 * Guest report for a single date.
 *
 * Assembles exactly what the printed document shows and nothing more. In
 * particular it **omits the amount**: this report is a guest register, and the
 * money side of a booking has no business being handed to whoever asks for it.
 * `stripPaymentInfo` enforces that at the type *and* the value level, so adding
 * a column to `customers` later cannot quietly leak it into the report.
 */

import { dayKeyOf, formatDate, toIso } from '../utils/date';
import { AppError } from '../utils/errors';
import type { BookingGuest, Customer } from '../types';
import type { BookingGuestRepository } from '../database/repositories/bookingGuestRepository';
import type { CustomerRepository } from '../database/repositories/customerRepository';
import type { SettingsService } from './settingsService';

/** A booking as it appears on the report: no payment fields, by construction. */
export interface ReportBooking {
  customer_code: string;
  name: string;
  address: string;
  phone: string;
  room_number: string;
  number_of_persons: number;
  check_in_date: string;
  check_out_date: string | null;
  id_front_path: string;
  id_back_path: string;
  id_front_thumb_path: string | null;
  id_back_thumb_path: string | null;
  /** Whether this booking was still open at the end of the reported day. */
  still_staying_on_date: boolean;
  arrived_on_date: boolean;
  departed_on_date: boolean;
  guests: ReportGuest[];
}

export interface ReportGuest {
  name: string;
  phone: string | null;
  id_front_path: string | null;
  id_back_path: string | null;
  id_front_thumb_path: string | null;
  id_back_thumb_path: string | null;
}

export interface DayReport {
  /** The reported local day, YYYY-MM-DD. */
  day: string;
  /** `23 Aug 2026`, ready to print. */
  day_label: string;
  hotel_name: string;
  hotel_address: string;
  hotel_phone: string;
  generated_at: string;
  bookings: ReportBooking[];
  totals: {
    bookings: number;
    /** Primary guests plus recorded companions. */
    people: number;
    arrivals: number;
    departures: number;
    rooms: number;
  };
}

const DAY_PATTERN = /^\d{4}-\d{2}-\d{2}$/;

/** `2026-08-23` -> local midnight on that date, for display only. */
function dayToLocalDate(day: string): Date {
  const [year, month, date] = day.split('-').map(Number);
  return new Date(year, month - 1, date);
}

/**
 * Copies only the fields the report is allowed to show.
 *
 * Written as an explicit whitelist rather than `delete row.amount_minor`, so a
 * future column is excluded by default instead of included by accident.
 */
export function stripPaymentInfo(
  customer: Customer,
  day: string,
  guests: BookingGuest[],
): ReportBooking {
  const checkOutDay = dayKeyOf(customer.check_out_date);
  return {
    customer_code: customer.customer_code,
    name: customer.name,
    address: customer.address,
    phone: customer.phone,
    room_number: customer.room_number,
    number_of_persons: customer.number_of_persons,
    check_in_date: customer.check_in_date,
    check_out_date: customer.check_out_date,
    id_front_path: customer.id_front_path,
    id_back_path: customer.id_back_path,
    id_front_thumb_path: customer.id_front_thumb_path,
    id_back_thumb_path: customer.id_back_thumb_path,
    arrived_on_date: dayKeyOf(customer.check_in_date) === day,
    departed_on_date: checkOutDay === day,
    still_staying_on_date: customer.check_out_date === null || checkOutDay > day,
    guests: guests.map((guest) => ({
      name: guest.name,
      phone: guest.phone,
      id_front_path: guest.id_front_path,
      id_back_path: guest.id_back_path,
      id_front_thumb_path: guest.id_front_thumb_path,
      id_back_thumb_path: guest.id_back_thumb_path,
    })),
  };
}

export class ReportService {
  constructor(
    private readonly customers: CustomerRepository,
    private readonly guests: BookingGuestRepository,
    private readonly settings: SettingsService,
  ) {}

  /**
   * Everyone whose stay covered `day` (YYYY-MM-DD), with their companions.
   */
  async buildDayReport(day: string, now: Date = new Date()): Promise<DayReport> {
    if (!DAY_PATTERN.test(day)) {
      throw new AppError('VALIDATION', 'Please choose a valid date for the report.');
    }

    const [rows, hotel] = await Promise.all([
      this.customers.listStayingOn(day),
      this.settings.get(),
    ]);

    // One query for every booking's companions rather than one per booking.
    const guestRows = await this.guests.listByCustomers(rows.map((r) => r.id));
    const guestsByCustomer = new Map<number, BookingGuest[]>();
    for (const guest of guestRows) {
      const list = guestsByCustomer.get(guest.customer_id);
      if (list) list.push(guest);
      else guestsByCustomer.set(guest.customer_id, [guest]);
    }

    const bookings = rows.map((row) =>
      stripPaymentInfo(row, day, guestsByCustomer.get(row.id) ?? []),
    );

    return {
      day,
      // Built from the parts as a LOCAL date: labelling via a UTC instant
      // would name the wrong day in far-eastern time zones.
      day_label: formatDate(toIso(dayToLocalDate(day))),
      hotel_name: hotel.hotel_name,
      hotel_address: hotel.hotel_address,
      hotel_phone: hotel.hotel_phone,
      generated_at: toIso(now),
      bookings,
      totals: {
        bookings: bookings.length,
        // The booked person count is the real headcount; recorded companions
        // are a subset of it, so counting both would double-count.
        people: bookings.reduce((sum, b) => sum + b.number_of_persons, 0),
        arrivals: bookings.filter((b) => b.arrived_on_date).length,
        departures: bookings.filter((b) => b.departed_on_date).length,
        rooms: new Set(bookings.map((b) => b.room_number)).size,
      },
    };
  }
}

File: lib/services/report_service.dart
import '../database/repositories/booking_guest_repository.dart';
import '../database/repositories/customer_repository.dart';
import '../models/models.dart';
import '../utils/app_date.dart';
import '../utils/app_error.dart';
import 'settings_service.dart';

/// A companion guest on a report — deliberately whitelisted, never carries
/// payment info (there is none on a guest anyway).
class ReportGuest {
  final String name;
  final String? phone;
  final String? idFrontPath;
  final String? idBackPath;
  const ReportGuest({required this.name, required this.phone, required this.idFrontPath, required this.idBackPath});
}

/// One booking's worth of report data. Deliberately excludes `amountMinor` —
/// this whitelist is what keeps payment amounts out of a document that may
/// be printed and shared, even if `Customer` gains new fields later.
class ReportBooking {
  final int customerId;
  final String customerCode;
  final String roomNumber;
  final String name;
  final String address;
  final String phone;
  final int numberOfPersons;
  final String? idFrontPath;
  final String? idBackPath;
  final String checkInDate;
  final String? checkOutDate;
  final bool arrivedOnDate;
  final bool departedOnDate;
  final bool stillStayingOnDate;
  final List<ReportGuest> guests;

  const ReportBooking({
    required this.customerId,
    required this.customerCode,
    required this.roomNumber,
    required this.name,
    required this.address,
    required this.phone,
    required this.numberOfPersons,
    required this.idFrontPath,
    required this.idBackPath,
    required this.checkInDate,
    required this.checkOutDate,
    required this.arrivedOnDate,
    required this.departedOnDate,
    required this.stillStayingOnDate,
    required this.guests,
  });
}

class ReportTotals {
  final int bookings;
  final int people;
  final int arrivals;
  final int departures;
  final int rooms;
  const ReportTotals({
    required this.bookings,
    required this.people,
    required this.arrivals,
    required this.departures,
    required this.rooms,
  });
}

class DayReport {
  final String day;
  final String dayLabel;
  final String hotelName;
  final String hotelAddress;
  final String hotelPhone;
  final String generatedAt;
  final List<ReportBooking> bookings;
  final ReportTotals totals;

  const DayReport({
    required this.day,
    required this.dayLabel,
    required this.hotelName,
    required this.hotelAddress,
    required this.hotelPhone,
    required this.generatedAt,
    required this.bookings,
    required this.totals,
  });
}

ReportBooking _stripPaymentInfo(Customer customer, String day, List<BookingGuest> guests) {
  final checkOutDay = customer.checkOutDate != null ? dayKeyOf(customer.checkOutDate!) : null;
  return ReportBooking(
    customerId: customer.id,
    customerCode: customer.customerCode,
    roomNumber: customer.roomNumber,
    name: customer.name,
    address: customer.address,
    phone: customer.phone,
    numberOfPersons: customer.numberOfPersons,
    idFrontPath: customer.idFrontPath,
    idBackPath: customer.idBackPath,
    checkInDate: customer.checkInDate,
    checkOutDate: customer.checkOutDate,
    arrivedOnDate: dayKeyOf(customer.checkInDate) == day,
    departedOnDate: checkOutDay == day,
    stillStayingOnDate: customer.checkOutDate == null || (checkOutDay != null && checkOutDay.compareTo(day) > 0),
    guests: guests
        .map((g) => ReportGuest(name: g.name, phone: g.phone, idFrontPath: g.idFrontPath, idBackPath: g.idBackPath))
        .toList(),
  );
}

final _dayPattern = RegExp(r'^\d{4}-\d{2}-\d{2}$');

class ReportService {
  final CustomerRepository _customers;
  final BookingGuestRepository _guests;
  final SettingsService _settings;
  ReportService(this._customers, this._guests, this._settings);

  Future<DayReport> buildDayReport(String day, {DateTime? now}) async {
    if (!_dayPattern.hasMatch(day)) {
      throw const AppError(AppErrorCode.validation, 'Enter a valid date.');
    }
    final results = await Future.wait([_customers.listStayingOn(day), _settings.get()]);
    final rows = results[0] as List<Customer>;
    final hotel = results[1] as HotelSettings;
    final guestsByCustomer = <int, List<BookingGuest>>{};
    if (rows.isNotEmpty) {
      final allGuests = await _guests.listByCustomers(rows.map((r) => r.id).toList());
      for (final guest in allGuests) {
        guestsByCustomer.putIfAbsent(guest.customerId, () => []).add(guest);
      }
    }
    final bookings = rows.map((r) => _stripPaymentInfo(r, day, guestsByCustomer[r.id] ?? const [])).toList();

    final dayDate = DateTime(int.parse(day.substring(0, 4)), int.parse(day.substring(5, 7)), int.parse(day.substring(8, 10)));
    return DayReport(
      day: day,
      dayLabel: formatDate(toIso(dayDate)),
      hotelName: hotel.hotelName,
      hotelAddress: hotel.hotelAddress,
      hotelPhone: hotel.hotelPhone,
      generatedAt: nowIso(),
      bookings: bookings,
      totals: ReportTotals(
        bookings: bookings.length,
        people: bookings.fold(0, (sum, b) => sum + b.numberOfPersons),
        arrivals: bookings.where((b) => b.arrivedOnDate).length,
        departures: bookings.where((b) => b.departedOnDate).length,
        rooms: bookings.map((b) => b.roomNumber).toSet().length,
      ),
    );
  }
}
